lib/std/io_uring.zig

```diff
@@ -0,0 +1,841 @@
+        if (count > 0) return count;
+        if (self.cq_ring_needs_flush() or wait_nr > 0) {
+            _ = try self.enter(0, wait_nr, linux.IORING_ENTER_GETEVENTS);
+            return self.copy_cqes_ready(cqes, wait_nr);
+        }
+        return 0;
+    }
+
+    fn copy_cqes_ready(self: *IO_Uring, cqes: []io_uring_cqe, wait_nr: u32) u32 {
+        const ready = self.cq_ready();
+        const count = std.math.min(cqes.len, ready);
+        const mask = self.cq.mask.*;
+        var head = self.cq.head.*;
+        var tail = head +% count;
+        // TODO Optimize this by using 1 or 2 memcpy's (if the tail wraps) rather than a loop.
+        var i: usize = 0;
+        // Do not use "less-than" operator since head and tail may wrap:
+        while (head != tail) {
+            cqes[i] = self.cq.cqes[head & mask]; // Copy struct by value.
+            head +%= 1;
+            i += 1;
+        }
+        self.cq_advance(count);
+        return count;
+    }
+
+    /// Returns a copy of an I/O completion, waiting for it if necessary, and advancing the CQ ring.
+    /// A convenience method for `copy_cqes()` for when you don't need to batch or peek.
+    pub fn copy_cqe(ring: *IO_Uring) !io_uring_cqe {
+        var cqes: [1]io_uring_cqe = undefined;
+        const count = try ring.copy_cqes(&cqes, 1);
+        assert(count == 1);
+        return cqes[0];
+    }
+
+    // Matches the implementation of cq_ring_needs_flush() in liburing.
+    fn cq_ring_needs_flush(self: *IO_Uring) bool {
+        return (@atomicLoad(u32, self.sq.flags, .Unordered) & IORING_SQ_CQ_OVERFLOW) > 0;
+    }
+
+    /// For advanced use cases only that implement custom completion queue methods.
+    /// If you use copy_cqes() or copy_cqe() you must not call cqe_seen() or cq_advance().
+    /// Must be called exactly once after a zero-copy CQE has been processed by your application.
+    /// Not idempotent, calling more than once will result in other CQEs being lost.
+    /// Matches the implementation of cqe_seen() in liburing.
+    pub fn cqe_seen(self: *IO_Uring, cqe: *io_uring_cqe) void {
+        self.cq_advance(1);
+    }
+
+    /// For advanced use cases only that implement custom completion queue methods.
+    /// Matches the implementation of cq_advance() in liburing.
+    pub fn cq_advance(self: *IO_Uring, count: u32) void {
+        if (count > 0) {
+            // Ensure the kernel only sees the new head value after the CQEs have been read.
+            @atomicStore(u32, self.cq.head, self.cq.head.* +% count, .Release);
+        }
+    }
+
+    /// Queues (but does not submit) an SQE to perform an `accept4(2)` on a socket.
+    /// Returns a pointer to the SQE.
+    pub fn queue_accept(
+        self: *IO_Uring,
+        user_data: u64,
+        fd: os.fd_t,
+        addr: *os.sockaddr,
+        addrlen: *os.socklen_t,
+        accept_flags: u32
+    ) !*io_uring_sqe {
+        if (!can_accept) return error.IO_UringKernelCannotAccept;
```

Comment:
Should this be a compile error instead?

Reply:
Thanks, like @daurnimator pointed out elsewhere, I think this needs to be a runtime check? Otherwise, someone could cross-compile this for Linux, and then someone else could take that and run it on an old kernel?

Reply:
This error depends on comptime condition, which depends on current build options. If you were to cross compile for newer kernel but run on older it would obviously be incorrect, its the same if you were to compile for the native target and then run on a generic cpu, you would get illegal instruction most probably.

Reply:
I think the logic for this constants should be something like this
```zig
const supported: enum {yes, no, check_at_runtime} = if (osMinVersion.isAtLeast(5, 4))
        .yes
    else if (osMaxVersion.isAtLeast(5, 4))
        .check_at_runtime
    else
        .no;
```
And so on for other constants.
Here you would do
```zig
if (can_accept == .no) {
    @compileError("unsupported accept");
} else if (can_accept == .check_at_runtime) {
    if (doRuntimeCheck()) return error.KernelDoesNotSupportAccept;
}
```

Reply:
Its better to always return an error. Users would probably have some fallback anyway and with comptime branch elimination there would no overhead. 
```zig
if (can_accept == .no or (can_accept == .check_at_runtime and self.doCheck())) {
    return error.UnsupportedKernel;
}